File: app/src/main/java/com/hover/stax/requests/NewRequestViewModel.java
package com.hover.stax.requests;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;

import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.contacts.StaxContact;
import com.hover.stax.schedules.Schedule;
import com.hover.stax.utils.DateUtils;
import com.hover.stax.transfers.AbstractFormViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class NewRequestViewModel extends AbstractFormViewModel {
	public final static String TAG = "NewRequestViewModel";

	private MediatorLiveData<Channel> activeChannel = new MediatorLiveData<>();

	private MutableLiveData<String> amount = new MutableLiveData<>();
	private MutableLiveData<List<StaxContact>> requestees = new MutableLiveData<>();
	private MediatorLiveData<String> requesterNumber = new MediatorLiveData<>();
	private MutableLiveData<String> note = new MutableLiveData<>();

	private MutableLiveData<Request> formulatedRequest = new MutableLiveData<>();
	private MutableLiveData<List<Request>> finalRequests = new MutableLiveData<>();

	public NewRequestViewModel(@NonNull Application application) {
		super(application);
		requestees.setValue(new ArrayList<>(Collections.singletonList(new StaxContact(""))));
		formulatedRequest.setValue(null);

		requesterNumber.addSource(activeChannel, this::setRequesterNumber);
	}

	void setAmount(String a) {
		amount.postValue(a);
	}

	LiveData<String> getAmount() {
		if (amount == null) {
			amount = new MutableLiveData<>();
		}
		return amount;
	}

	void setActiveChannel(Channel c) {
		activeChannel.setValue(c);
	}
	LiveData<Channel> getActiveChannel() {
		return activeChannel;
	}

	void setRequesterNumber(Channel c) {
		if (c != null) requesterNumber.setValue(c.accountNo);
	}

	void setRequesterNumber(String number) {
		requesterNumber.postValue(number);
		if (activeChannel.getValue() != null)
			activeChannel.getValue().accountNo = number;
	}
	LiveData<String> getRequesterNumber() {
		if (requesterNumber == null) { requesterNumber = new MediatorLiveData<>(); }
		return requesterNumber;
	}

	public void onUpdate(int pos, StaxContact contact) {
		List<StaxContact> cs = requestees.getValue() != null ? requestees.getValue() : new ArrayList<>();
		try { cs.set(pos, contact);
		} catch (IndexOutOfBoundsException e) { cs.add(contact); }
		requestees.postValue(cs);
	}

	void addRecipient(StaxContact contact) {
		List<StaxContact> rList = requestees.getValue() != null ? requestees.getValue() : new ArrayList<>();
		rList.add(contact);
		requestees.postValue(rList);
	}

	LiveData<List<StaxContact>> getRequestees() {
		if (requestees == null) {
			requestees = new MutableLiveData<>();
		}
		return requestees;
	}

	void resetRecipients() { requestees.setValue(new ArrayList<>()); }

	void setNote(String n) {
		note.postValue(n);
	}

	LiveData<String> getNote() {
		if (note == null) {
			note = new MutableLiveData<>();
			note.setValue(" ");
		}
		return note;
	}

	boolean validAmount() { return (amount.getValue() != null && !amount.getValue().isEmpty() && amount.getValue().matches("\\d+") && !amount.getValue().matches("[0]+")); }

	String requesteeErrors() {
		if (requestees.getValue() != null && requestees.getValue().size() > 0 && !requestees.getValue().get(0).getPhoneNumber().isEmpty())
			return null;
		return getApplication().getString(R.string.request_error_recipient);
	}

	String requesterAcctNoError() {
		if (requesterNumber.getValue() == null && !requesterNumber.getValue().isEmpty())
			return null;
		return getApplication().getString(R.string.requester_number_fielderror);
	}

	boolean validNote() { return (note.getValue() != null && !note.getValue().isEmpty()); }

	public void setSchedule(Schedule s) {
		schedule.setValue(s);
		setAmount(s.amount);
		new Thread(() -> requestees.postValue(repo.getContacts(s.recipient_ids.split(",")))).start();
		setNote(s.note);
	}

	public void createRequest() {
		repo.update(activeChannel.getValue());
		saveContacts();
		formulatedRequest.setValue(new Request(amount.getValue(), note.getValue(), requesterNumber.getValue(), activeChannel.getValue().institutionId));
	}

	LiveData<Request> getRequest() {
		if (formulatedRequest == null) { formulatedRequest = new MutableLiveData<>(); }
		return formulatedRequest;
	}

	void saveRequest() {
		if (finalRequests.getValue() != null && finalRequests.getValue().size() > 0 && requestees.getValue() != null && finalRequests.getValue().size() == requestees.getValue().size())
			return;
		ArrayList<Request> requests = new ArrayList<>();
		for (StaxContact recipient : requestees.getValue()) {
			Request r = new Request(formulatedRequest.getValue(), recipient, getApplication());
			requests.add(r);
			repo.insert(r);
		}
		finalRequests.setValue(requests.size() > 0 ? requests : null);
	}

	LiveData<List<Request>> getRequests() {
		if (finalRequests == null) { finalRequests = new MutableLiveData<>(); }
		return finalRequests;
	}

	void removeInvalidRequestees() {
		if (requestees.getValue() != null && requestees.getValue().size() > 0) {
			List<StaxContact> contacts = new ArrayList<>();
			for (StaxContact c: requestees.getValue()) {
				if (c.phoneNumber != null && !c.phoneNumber.isEmpty())
					contacts.add(c);
			}
			requestees.setValue(contacts);
		}
	}

	public void saveContacts() {
		if (requestees.getValue() != null) {
			new Thread(() -> {
				for (StaxContact c: requestees.getValue()) {
					c.lastUsedTimestamp = DateUtils.now();
					repo.insertOrUpdate(c);
				}
			}).start();
		}
	}
}

File: app/src/main/java/com/hover/stax/requests/NewRequestViewModel.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.requests

import android.app.Application
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.hover.stax.R
import com.hover.stax.contacts.ContactRepo
import com.hover.stax.contacts.StaxContact
import com.hover.stax.data.local.accounts.AccountRepo
import com.hover.stax.domain.model.Account
import com.hover.stax.schedules.Schedule
import com.hover.stax.schedules.ScheduleRepo
import com.hover.stax.transfers.AbstractFormViewModel
import com.hover.stax.utils.DateUtils
import java.util.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class NewRequestViewModel(
    application: Application,
    val repo: RequestRepo,
    val accountRepo: AccountRepo,
    contactRepo: ContactRepo,
    scheduleRepo: ScheduleRepo
) : AbstractFormViewModel(application, contactRepo, scheduleRepo) {

    val activeAccount = MutableLiveData<Account?>()
    val amount = MutableLiveData<String?>()
    private val requestees = MutableLiveData<List<StaxContact>>(Collections.singletonList(StaxContact("")))
    val requestee = MutableLiveData<StaxContact?>()
    var requesterNumber = MediatorLiveData<String>()
    val note = MutableLiveData<String?>()

    val formulatedRequest = MutableLiveData<Request>()
    private val finalRequests = MutableLiveData<List<Request>>()

    init {
        requesterNumber.addSource(activeAccount) { setRequesterNumber(it) }
    }

    fun setAmount(a: String?) = amount.postValue(a)

    fun setActiveAccount(account: Account) = activeAccount.postValue(account)

    private fun setRequesterNumber(a: Account?) {
        requesterNumber.postValue(a?.accountNo ?: "")
    }

    fun setRequesterNumber(number: String) {
        activeAccount.value?.let { it.accountNo = number }
        requesterNumber.postValue(number)
    }

    fun setRecipient(recipient: String) {
        if (requestee.value != null && requestee.value.toString() == recipient) return
        requestee.value = StaxContact(recipient)
    }

    fun addRecipient(contact: StaxContact) {
        requestee.value = contact
    }

    fun setNote(n: String?) = note.postValue(n)

    fun validAmount(): Boolean = (!amount.value.isNullOrEmpty() && amount.value!!.matches("\\d+".toRegex()) && !amount.value!!.matches("[0]+".toRegex()))

    fun requesteeErrors(): String? {
        return if (!requestee.value?.accountNumber.isNullOrEmpty())
            null
        else
            getString(R.string.request_error_recipient)
    }

    fun accountError(): String? = if (activeAccount.value != null) null else getString(R.string.accounts_error_noselect)

    fun requesterAcctNoError(): String? = if (!requesterNumber.value.isNullOrEmpty()) null else getString(R.string.requester_number_fielderror)

    fun validNote(): Boolean = !note.value.isNullOrEmpty()

    // TODO validate that this works from schedule
    fun setSchedule(s: Schedule) {
        schedule.postValue(s)
        setAmount(s.amount)

        viewModelScope.launch {
            val contacts = contactRepo.getContacts(s.recipient_ids.split(",").toTypedArray())
            requestees.postValue(contacts)
        }
    }

    fun createRequest() {
        saveContacts()

        activeAccount.value?.institutionId?.let {
            val request = Request(amount.value, note.value, requesterNumber.value, it)
            formulatedRequest.value = request
        }
    }

    fun saveRequest() {
        if (formulatedRequest.value != null) {
            viewModelScope.launch(Dispatchers.IO) {
                val r = Request(formulatedRequest.value!!, requestee.value, getApplication())
                repo.insert(r)

                finalRequests.postValue(listOf(r))
            }
        }
    }

    private fun saveContacts() {
        requestee.value?.let { contact ->
            viewModelScope.launch {
                if (!contact.accountNumber.isNullOrEmpty()) {
                    contact.lastUsedTimestamp = DateUtils.now()
                    contactRepo.save(contact)
                }
            }
        }
    }

    override fun reset() {
        super.reset()
        setAmount(null)
        setNote(null)
        requestee.value = null
    }
}